part/txn.go

```diff
@@ -0,0 +1,347 @@
+			}
+			newNode.setKind(nodeKind4)
+			newNode.setSize(2)
+			if !txn.opts.rootOnlyWatch || this == newRoot {
+				newNode.node.watch = make(chan struct{})
+			}
+			*thisp = newNode.self()
+			return
+		}
+	}
+}
+
+type deleteParent[T any] struct {
+	node  *node[T]
+	index int
+}
+
+func (txn *Txn[T]) delete(root *node[T], key []byte) (oldValue T, hadOld bool, newRoot *node[T]) {
+	newRoot = root
+	this := root
+
+	// Reuse the same slice in the transaction to hold the parents in order to avoid
+	// allocations.
+	if txn.deleteParents == nil {
+		txn.deleteParents = make([]deleteParent[T], 0, 32)
+	}
+	parents := txn.deleteParents[:0]
+	parents = append(parents, deleteParent[T]{nil, -1}) // Placeholder for new root.
+
+	// Find the target node and record the path to it.
+	for {
+		if bytes.HasPrefix(key, this.prefix) {
+			key = key[len(this.prefix):]
+			if len(key) == 0 {
+				if this.leaf == nil {
+					// Not found.
+					return
+				}
+				// Target node found!
+				break
+			}
+			var idx int
+			this, idx = this.findIndex(key[0])
+			if this == nil {
+				return
+			}
+			parents = append(parents, deleteParent[T]{this, idx})
+		} else {
+			// Reached a node with a different prefix, so node not found.
+			return
+		}
+	}
+
+	oldValue = this.leaf.value
+	hadOld = true
+
+	// The target was found, rebuild the tree from the root upwards.
+	newRoot = txn.cloneNode(root)
+	parents[0].node = newRoot
+
+	if this == root {
+		// Target is the root, clear it.
+		newRoot.leaf = nil
+		if newRoot.size() == 0 {
+			// No children so we can clear the prefix.
+			newRoot.prefix = nil
+		}
+		return
+	}
+
+	for i := len(parents) - 1; i > 0; i-- {
+		parent := &parents[i-1]
+		target := &parents[i]
+
+		// Clone the parent to mutate it.
+		parent.node = txn.cloneNode(parent.node)
+		var children []*node[T]
+		if parent.node.kind() == nodeKind256 {
+			children = parent.node.node256().children[:]
+		} else {
+			children = parent.node.children()
+		}
+
+		if target.node == this && target.node.size() > 0 {
+			// This is the node that we want to delete, but it has
+			// children. Clone and clear the leaf.
+			target.node = txn.cloneNode(target.node)
+			target.node.leaf = nil
+			children[target.index] = target.node
+		}
+
+		if target.node.size() == 0 && (target.node == this || target.node.leaf == nil) {
+			// The node is empty, remove it from the parent.
+			if parent.node.kind() == nodeKind256 {
+				children[target.index] = nil
+			} else {
+				copy(children[target.index:], children[target.index+1:])
+				children[parent.node.size()-1] = nil
+			}
+			parent.node.setSize(parent.node.size() - 1)
+		} else {
+			// Update target to point to the cloned node
+			children[target.index] = target.node
+		}
+
+		if parent.node.size() > 0 {
+			// Check if the node should be demoted.
+			// TODO: Can we avoid the initial clone of parent.node?
+			var newNode *node[T]
+			switch {
+			case parent.node.kind() == nodeKind256 && parent.node.size() <= 48:
```

Comment:
If we don't demote at the border we would be wasting space and then we might as well make the node a different size?

How would the underlying array be retained (it's embedded in the node) and wouldn't it be bad to retain it since we would then use more memory?

Reply:
I set the demoting to be at 37, 12 and 3 to avoid the thrashing.